lib/ansible/plugins/inventory/scaleway.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: scaleway
+    plugin_type: inventory
+    authors:
+      - Remy Leone <rleone@online.net>
+    short_description: Scaleway inventory source
+    description:
+        - Get inventory hosts from Scaleway
+'''
+
+EXAMPLES = '''
+# scaleway_inventory.yml file in YAML format
+# Example command line: ansible-inventory --list -i scaleway_inventory.yml
+
+plugin: scaleway
+regions:
+  - ams1
+  - par1
+tags:
+  - foobar
+'''
+
+import json
+import os
+
+from ansible.errors import AnsibleError
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.module_utils.scaleway import SCALEWAY_LOCATION
+from ansible.module_utils.urls import open_url
+
+
+def _fetch_information(token, url):
+    try:
+        response = open_url(url,
+                            headers={'X-Auth-Token': token,
+                                     'Content-type': 'application/json'})
+    except Exception:
+        raise AnsibleError("Error while fetching %s" % url)
+
+    try:
+        raw_json = json.loads(response.read())
+    except ValueError:
+        raise AnsibleError("Incorrect JSON payload")
+
+    try:
+        return raw_json["servers"]
+    except KeyError:
+        raise AnsibleError("Incorrect format from the Scaleway API response")
+
+
+def _build_server_url(api_endpoint):
+    return "/".join([api_endpoint, "servers"])
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'scaleway'
+
+    def __init__(self):
+        super(InventoryModule, self).__init__()
+
+        self.token = os.environ["SCW_TOKEN"]
+        self.config_data = None
+
+    def verify_file(self, path):
+        return "scaleway" in path
+
+    def _fill_host_variables(self, server_id, server_info):
+        targeted_attributes = (
+            "arch",
+            "commercial_type",
+            "organization",
+            "state",
+            "hostname",
+            "state"
+        )
+        for attribute in targeted_attributes:
+            self.inventory.set_variable(server_id, attribute, server_info[attribute])
+
+        self.inventory.set_variable(server_id, "tags", server_info["tags"])
+        self.inventory.set_variable(server_id, "ipv4", server_info["public_ip"]["address"])
+
+    @property
+    def zones(self):
+        config_zones = self.config_data.get("regions", SCALEWAY_LOCATION.keys())
+        return set(SCALEWAY_LOCATION.keys()).intersection(config_zones)
+
+    @property
+    def tags(self):
+        return self.config_data.get("tags", None)
```

Comment:
See other plugins for how to use set_options/get_option base methods, these not only work from documentation but they also can add validation and defaults.

Reply:
I've renamed them to `_get_zones` and `_get_tags` is there additional steps I can do to improve my code?

Reply:
not what i meant, dont access config_data directly, using `get_option()` will verify type and ensure you get the default you set.

Reply:
Done tell me if it's ok now.